src/GraphQL/Types/Collections/SchemaTypes.cs

```diff
@@ -189,25 +189,25 @@ protected void Initialize(ISchema schema, IServiceProvider serviceProvider, IEnu
                 {
                     if (this[name] == null)
                     {
-                        AddType(graphType, context);
+                        AddType(graphType, context, schema);
                     }
                 },
                 graphTypeMappings);
 
-            foreach (var type in types)
+            foreach (var type in types.Distinct())
             {
                 if (type is not ScalarGraphType)
-                    AddType(type, ctx);
+                    AddTypeIfNotRegistered(type, ctx, schema);
             }
 
             // these fields must not have their field names translated by INameConverter; see HandleField
-            HandleField(null, SchemaMetaFieldType, ctx, false);
-            HandleField(null, TypeMetaFieldType, ctx, false);
-            HandleField(null, TypeNameMetaFieldType, ctx, false);
+            HandleField(null, SchemaMetaFieldType, ctx, false, schema);
```

Comment:
Maybe add `Schema` into `TypeCollectionContext`?

Reply:
I thought of it, but the context is recreated a couple times in the initialization code.  I could put it in the context of course.

(Personally I think the entire class needs a rewrite, but it's more work than I've wanted to do so far.)

Reply:
ok done